### ISSUE TITLE
`// ORDERING: ...` comment for atomic operations

### ISSUE THREAD

**Issue by ojeda** [rust-lang/rust-clippy#16073]
### What it does

Similar to `// SAFETY: ...`, `// PANIC: ...` and `// CAST: ...` comments, but for atomic operations that use one or more `Ordering`s.

Note that in the Linux kernel we use a custom implementation of atomics, so we should probably pass a list of functions/methods/... to warn about, similar to `// PANIC: ...`.

More generally, we could consider an implementation that warns on functions/methods that take `Ordering` parameters unless opt-ed out. But just the list as for `// PANIC: ...` sounds good enough, and perhaps it makes sense to instead generalize that lint to allow the user to customize a map of `TAG`-> list of methods that require a comment.

That is, we would like code to look like:

```rust
// ORDERING: Relaxed is fine because we don't expect synchronization here.
let old = self.init.xchg(1, Relaxed);
```

In other words, like `undocumented_unsafe_blocks`, but for atomic operations.

In addition, there would be a dual lint to detect unnecessary comments: https://github.com/rust-lang/rust-clippy/issues/16074.

Cc: @blyxyas @hcbarker @fbq

### Advantage

The advantages are very similar to the ones we have seen from applying the `// SAFETY: ...` convention:

  - Atomic operations get documented better, i.e. the reason a particular `Ordering` is used must be explained.

  - It gives some pause to developers when picking an `Ordering`.

These advantages also mean review time (and thus maintainers' workload) gets reduced, since reviewers will have an easier time following the rationale behind the `Ordering`s picked.

### Example

```rust
let old = self.init.xchg(1, Relaxed);
```

Should be written as:

```rust
// ORDERING: Relaxed is fine because we don't expect synchronization here.
let old = self.init.xchg(1, Relaxed);
```

### Additional Context

Please see the "Additional context" for `// PANIC: ...` on https://github.com/rust-lang/rust-clippy/issues/15895.

**Comment by y21** [comment]
I'm personally not too much of a fan of these tag comments, given that `#[expect]` achieves basically the same thing. I'm aware that there's some arguments for them in #15895 but they don't seem convincing enough for me for the complexity this would introduce, the dual lint #16074 for example is also simply not needed if this were just a normal restriction lint that people `#[expect]`, since that already warns if the lint doesn't emit there.

> Or, seen from another perspective: these // PANIC: comments are not so much about allowing a false positive of a lint, but rather about writing a required comment.

I'd argue most restriction (or even some pedantic lints, like the various cast ones) are intended to be used exactly like this; in many instances the pattern to lint against is not a false positive and you want to `#[expect]` them. Combined with `allow_attributes_without_reason` it's also possible to require a comment.

**Comment by blyxyas** [comment]
I talked with Miguel about this concern in Kangrejos this year, and he denoted the aspect that `#[expect]` is much more technical, people see it more of an ad-hoc justification of why the "bad pattern is permitted". In these cases, the `// PANIC`, "// CAST", etc... comments move from a "ignore this warning, I know it's bad" mentality into a "I know what I'm doing, and this is the reason" mentality.

That's the reason I don't mind having two concepts that technically do the same, but have vastly different ways of thinking.

(Also, I think that the Rust-skeptic crowd doesn't like to see `allow`s for our lints)

**Comment by ojeda** [comment]
@y21 Attributes are a quite bad fit for this, in my opinion.

They are harder to read and write, they seem worse for multiline, they require providing a lint level which is not relevant (`expect`), they require a lint name that may be longer than needed (e.g. a `clippy::` prefix and possibly more, e.g. a suffix) and that name may be an implementation detail after all.

Really, what we want is an annotation that explains something about the code that follows. Which is what a code comment is, after all.

In my view, it is quite clear just by looking at it:

```rust
#[expect(clippy::ordering_tag, reason = "Relaxed is fine because we don't expect synchronization here.")]
let old = self.init.xchg(1, Relaxed);
```

```rust
// ORDERING: Relaxed is fine because we don't expect synchronization here.
let old = self.init.xchg(1, Relaxed);
```

It is not just about the length, but about the fact that, with the attribute approach, the most important information is mixed with other stuff in the middle of the line (in the case of the tag kind) and at the end of the line (in the case of the reason).

Moreover, using a lint attribute for this is, in my view, abusing the lint system. For instance, since we support several compiler versions, we need to support ignoring only those unknown `expect`s without wrapping with `cfg`s. Maybe there is a way for that already, but my point is that such a problem may appear because it was not meant for that. Similarly, we would need a way to customize `allow_attributes_without_reason` for certain lints only.

There are other annoying bits when considering the wider context, e.g. it will be inconsistent with existing `// SAFETY: ...` comments and, when one sees an attribute, one will need to scan it to see the lint name just to understand if it is an explanation (i.e. like a tag) or about a false positive (i.e. like a lint attribute).

> I'd argue most restriction (or even some pedantic lints, like the various cast ones) are intended to be used exactly like this;

I don't know if most (quite a few look to me like about replacing/avoiding the pattern completely and I don't see why one would use `#[expect]`), but it is true that some may be used as you say. For those, I think a tag would have been best for the reasons mentioned above.

> for the complexity this would introduce

I would imagine most of these tags would work similarly to other lints, especially since nowadays the infrastructure for `// SAFETY: ...` is there.

**Comment by felix91gr** [comment]
Hi!

I'm thinking that for the work of the SCRC we'd probably want something like this (i.e. to have a way to ask for documentation as for the reasoning of the `Ordering` of atomic ops, in the same way as the `undocumented_unsafe_blocks` lint allows to do).

@ojeda, regarding the format (attribute v/s comment):

My tendency in these things is to prefer something with more semantic structure than just a comment; if only because that tends to integrate better with tooling, specially at scale.

Now, if we know beforehand that all the structure we will need for this is only of the form of
```rust
// ORDERING: ...
```
, followed by 0+ line comments, then I think the structure is simple enough that all tools would integrate quite well with it (e.g. things like `ODRERING` would trigger a lint with no false positives nor false negatives, and a mouse-over action on the IDE could explain to you exactly why this kind of comment is there).

But if the structure were to be only a tiny bit more complex, then it could start causing trouble. Lesser tools would have a hard time reasoning about it, and things like LSP servers might be limited in how they can help the user.

So in that sense, leveraging attributes would be a lot better. The entire support infrastructure for this kind of annotation is there already.

---

Aside:

I have a random thought, which is that perhaps a third kind of annotation might be even better. Something that's not an attribute nor a comment; something designed to do exactly this kind of thing.

But I bet that's not viable at the lang level. I dunno... but sometimes when none of the two options seem great, a third, independent option, can be better.

(I bet it's very non-viable)

**Comment by ojeda** [comment]
It would all depend on the structure one is expecting. So far, this has been about explaining code below, and using English for it. If it is about something else, then yeah, it could require a different system.

In any case, the current lint system only gives us a string anyway as far as I understand, and thus I guess you are talking about new attributes altogether.

And if it is a new kind of annotation altogether, then that could be potentially nice, depending on what it is, but should still be something easy to read and to type for humans.

For instance, for something like contracts, I would very much prefer to avoid attributes (like some existing tooling uses) and use new syntax with e.g. `pre` and `post` keywords similar to how `where` looks like.

But for anything that resembles mostly free-form English and is about explaining what code does, then a simple comment-like thing seems best to me.
